### ISSUE TITLE
`best_inference_error` and `best_val_loss` can be lost during training restart

### ISSUE THREAD

**Issue by spencerkclark** [ai2cm/ace#651]
In a recent training job ([beaker](https://beaker.allen.ai/orgs/ai2/workspaces/ace/work/01K0HQBQ45YXV953XJ1SHK2FZ6?taskId=01K0HQBQ4MG1WMFBSSK607STA8&jobId=01K0Q8ATETNXXMY2AC979VW6DZ), [WandB](https://wandb.ai/ai2cm/ace-som/runs/vwob168g)), which was pre-empted several times, I noticed that the `best_inference_error` was occasionally reset to infinity, which could interfere with best inference checkpoint selection.  Could this be a side effect of the recent checkpointing during epochs update?

<p align="center">
<img width="50%" height="50%" alt="Image" src="https://github.com/user-attachments/assets/2d21e614-2982-453f-b205-923f8d095dac" />
<p align="center">

As @mcgibbon notes, the same is true for `best_val_loss`:

<p align="center">
<img width="50%" height="50%"  alt="Image" src="https://github.com/user-attachments/assets/26c489a7-47f2-47c9-a2aa-df1c19940963" />
</p>

**Comment by spencerkclark** [comment]
It seems like this may still be an issue after ai2cm/full-model#2037, e.g. [this WandB training job](https://wandb.ai/ai2cm/ace-som/runs/9jd06qto) was done with that change and still exhibits reset `best_val_loss` and `best_inference_error` during training. 

Something else I have observed is that both of these are logged as infinite in the first epoch, regardless of whether the job has been pre-empted or not.  I think this may be because these metrics are logged prior to being updated in `self.save_all_checkpoints`: https://github.com/ai2cm/full-model/blob/46100aafc1e119573b1c250e2f6a3e7379e9240f/fme/core/generics/trainer.py#L356-L380
Before ai2cm/full-model#2019 this logic occurred in reverse, so `self._best_validation_loss` and `self._best_inference_error` would be updated appropriately based on the current epoch's `valid_loss` and `inference_error` prior to being logged.

**Comment by mcgibbon** [comment]
It looks like the model was [pre-empted during validation](https://beaker.allen.ai/orgs/ai2/workspaces/climate-titan/work/01K0ZJ2Y2ZR81CG01C14K66RCS/logs?jobId=01K0ZJ43RA2FD4XJG780J73H9R) on epoch 52 (where the blip happens). During this epoch, no on-termination checkpoint got saved because the exception was tracked through a failing data worker (which I should fix). Then it was started again, over-wrote the checkpoint before starting validation, and was [pre-empted during validation again](https://beaker.allen.ai/orgs/ai2/workspaces/climate-titan/work/01K0ZJ2Y2ZR81CG01C14K66RCS/logs?jobId=01K11B8M2FS4EQQCWZZSMYBVT7). Then it [completed validation/inference](https://beaker.allen.ai/orgs/ai2/workspaces/climate-titan/work/01K0ZJ2Y2ZR81CG01C14K66RCS/logs?jobId=01K11DNH6FHF5CD79MVRE4Z1KQ) for that epoch, reporting a previous val_loss of inf. Sample 56992 is the last pre-blip validation loss, 58088 is where the inf occurs (end of epoch after 52 completed epochs), and then 59184 is the next non-inf validation loss - these line up with the values I see in the logs.

This means the issue is happening even without the on-termination call, which was never triggered in this run.

**Comment by mcgibbon** [comment]
Looking now at [this run](https://beaker.allen.ai/orgs/ai2/workspaces/climate-titan/work/01K0ZJ350NHBKQHRDAFZNQJHTB?taskId=01K0ZJ3514CEVW7W8M2V7S0JMH&jobId=01K18WAQA0DNEQ6EQ3VV58WJCB) which has some interesting epochs. after the epoch ending at 206048, the epoch at 207144 has higher best validation loss, and lower train time suggests it got pre-empted during the training. However, it keeps the same best inference loss. Run 5 shows a checkpoint saved at the end of epoch after 207144 and then pre-empted during inference. [Run 6](https://beaker.allen.ai/api/job/01K18WAQA0DNEQ6EQ3VV58WJCB/logs/all) shows that data as loaded. So even on a simple pre-emption during inference, we are seeing this error, and it can happen in only the validation loss.

**Comment by spencerkclark** [comment]
ai2cm/full-model#2057 and ai2cm/full-model#2058 were incorporated into training jobs I started on Friday on Ursa, and they all seem to be behaving as expected in this respect ([WandB view](https://wandb.ai/ai2cm/ace-som?nw=1nsmgvoosqk)), so I think it is safe to say this issue is resolved. The jobs are set to stop and requeue every three hours, so in some sense are a test of jobs that are frequently pre-empted.  Thanks again @mcgibbon!
